Avoid duplicating option metadata across daemons

nrwahl2 opened the pull request:
Closes [T746](https://projects.clusterlabs.org/T746)

kgaillot commented on lib/common/nvpair.c:
```diff
@@ -871,7 +871,7 @@ xml2list(const xmlNode *parent)
 pcmk__xe_set_bool_attr(xmlNodePtr node, const char *name, bool value)
 {
-    crm_xml_add(node, name, value ? XML_BOOLEAN_TRUE : XML_BOOLEAN_FALSE);
+    crm_xml_add(node, name, value ? PCMK_VALUE_TRUE : XML_BOOLEAN_FALSE);
```
This can be `pcmk__btoa(value)`

nrwahl2 replied:
Thanks

kgaillot commented on lib/common/tests/options/pcmk__env_option_enabled_test.c:
```diff
@@ -27,11 +27,11 @@ enabled_true_value(void **state)
 
     expect_string(__wrap_getenv, name, "PCMK_env_var");
-    will_return(__wrap_getenv, "true");
+    will_return(__wrap_getenv, PCMK_VALUE_TRUE);
```
This test should not be changed, since it is a test that the literal string "true" is treated as true (effectively, it is also testing that PCMK_VALUE_TRUE is defined correctly)

kgaillot commented on lib/common/tests/strings/crm_is_true_test.c:
```diff
@@ -18,7 +18,7 @@ bad_input(void **state) {
 static void
 is_true(void **state) {
-    assert_true(crm_is_true("true"));
+    assert_true(crm_is_true(PCMK_VALUE_TRUE));
```
ditto

kgaillot commented on lib/common/tests/strings/crm_str_to_boolean_test.c:
```diff
@@ -42,7 +42,7 @@ is_true(void **state) {
 
     // Ensure it still validates the string with a NULL result argument
-    assert_int_equal(crm_str_to_boolean("true", NULL), 1);
+    assert_int_equal(crm_str_to_boolean(PCMK_VALUE_TRUE, NULL), 1);
```
ditto

kgaillot commented on lib/common/tests/strings/pcmk__btoa_test.c:
```diff
@@ -14,7 +14,7 @@
 btoa(void **state) {
     assert_string_equal(pcmk__btoa(false), "false");
-    assert_string_equal(pcmk__btoa(true), "true");
+    assert_string_equal(pcmk__btoa(true), PCMK_VALUE_TRUE);
```
ditto

kgaillot commented on lib/common/tests/options/pcmk__env_option_enabled_test.c:
```diff
@@ -44,11 +44,11 @@ disabled_false_value(void **state)
 
     expect_string(__wrap_getenv, name, "PCMK_env_var");
-    will_return(__wrap_getenv, "false");
+    will_return(__wrap_getenv, PCMK_VALUE_FALSE);
```
ditto

kgaillot commented on lib/common/tests/strings/crm_is_true_test.c:
```diff
@@ -31,7 +31,7 @@ is_true(void **state) {
 static void
 is_false(void **state) {
-    assert_false(crm_is_true("false"));
+    assert_false(crm_is_true(PCMK_VALUE_FALSE));
```
ditto

kgaillot commented on lib/common/tests/strings/crm_str_to_boolean_test.c:
```diff
@@ -82,7 +82,7 @@ is_false(void **state) {
 
     // Ensure it still validates the string with a NULL result argument
-    assert_int_equal(crm_str_to_boolean("false", NULL), 1);
+    assert_int_equal(crm_str_to_boolean(PCMK_VALUE_FALSE, NULL), 1);
```
ditto

kgaillot commented on lib/common/tests/strings/pcmk__btoa_test.c:
```diff
@@ -13,9 +13,9 @@
+    assert_string_equal(pcmk__btoa(false), PCMK_VALUE_FALSE);
     assert_string_equal(pcmk__btoa(true), PCMK_VALUE_TRUE);
-    assert_string_equal(pcmk__btoa(1 == 0), "false");
+    assert_string_equal(pcmk__btoa(1 == 0), PCMK_VALUE_FALSE);
```
ditto

kgaillot reviewed (commented):
Reviewed through "Deprecate XML_BOOLEAN_FALSE", let's do a separate PR through that. Check copyrights.

kgaillot commented:
For stonith-watchdog-timeout, I'd say just validate it as a timeout value in the option table, then do extra validation in the controller when needed (I guess at startup), then drop the cleanup stuff

nrwahl2 commented on daemons/controld/controld_messages.c:
```diff
@@ -1119,12 +1117,6 @@ handle_request(xmlNode *stored_msg, enum crmd_fsa_cause cause)
-        crm_shutdown(SIGTERM);
-        /*return I_SHUTDOWN; */
-        return I_NULL;
-
```
Ah, makes sense

nrwahl2 commented:
> Reviewed through "Deprecate XML_BOOLEAN_FALSE", let's do a separate PR through that. Check copyrights.

I meant to mention this in the PR description. I pulled those four commits from one of my symbol renaming branches, just so that the merge conflict won't be as huge (since big arrays are moving around between files).

nrwahl2 commented on daemons/fenced/pacemaker-fenced.c:
```diff
@@ -530,184 +530,261 @@ static pcmk__cluster_option_t fencer_options[] = {
+        PCMK_STONITH_HOST_LIST, NULL, "string", NULL,
+        "", NULL,
+        N_("A list of machines controlled by this device (optional unless "
+            "pcmk_host_list=\"static-list\")"),
```
Thanks, and ditto below

nrwahl2 commented on lib/pengine/tests/status/set_working_set_defaults_test.c:
```diff
@@ -28,7 +28,7 @@ check_defaults(void **state) {
             |pcmk_sched_stop_removed_resources
             |pcmk_sched_cancel_removed_actions;
 
```
Okay in this context?

nrwahl2 commented on lib/pengine/tests/status/set_working_set_defaults_test.c:
```diff
@@ -28,7 +28,7 @@ check_defaults(void **state) {
             |pcmk_sched_cancel_removed_actions;
 
-    if (!strcmp(PCMK__CONCURRENT_FENCING_DEFAULT, "true")) {
+    if (!strcmp(PCMK__CONCURRENT_FENCING_DEFAULT, PCMK_VALUE_TRUE)) {
```
Okay in this context?

kgaillot replied:
Yes, because we're not comparing user input, we're comparing our own settings (via configure in this case)

nrwahl2 commented on lib/common/tests/nvpair/pcmk__xe_set_bool_attr_test.c:
```diff
@@ -21,7 +21,7 @@ set_attr(void **state)
     pcmk__xe_set_bool_attr(node, "b", false);
 
-    assert_string_equal(crm_element_value(node, "a"), XML_BOOLEAN_TRUE);
+    assert_string_equal(crm_element_value(node, "a"), PCMK_VALUE_TRUE);
```
These tests already used the constants. Is that still okay since we're testing at a slightly higher level (we're effectively testing the value of `PCMK_VALUE_TRUE` elsewhere), or switch to literals?

kgaillot replied:
This is fine too, again because our code set the value to be compared against (technically the other tests are our code themselves, but they're simulating an external caller)

nrwahl2 commented:
The last few commits are preparation to create real XML (instead of just an XML string) for XML formatted output.

In hindsight `string2xml()` might have worked, but the constants need to get created anyway. Might as well do it now.

kgaillot commented:
> Alternatively, it occurred to me that we could call `pcmk_controld_api_node_info()` from within the timeout validator to get the node name. We'd only have to do it on the first call, and then store the node name for later reuse. That way we could validate the watchdog timeout from all daemons and get a config warning in the scheduler.

We don't want to make any daemon connections in option validation. It could be called from a command-line tool on a saved file without the cluster running.

kgaillot commented:
> > Alternatively, it occurred to me that we could call `pcmk_controld_api_node_info()` from within the timeout validator to get the node name. We'd only have to do it on the first call, and then store the node name for later reuse. That way we could validate the watchdog timeout from all daemons and get a config warning in the scheduler.
> 
> We don't want to make any daemon connections in option validation. It could be called from a command-line tool on a saved file without the cluster running.

And pacemaker-schedulerd shouldn't do it itself either, because we want crm_simulate to get identical results (and the daemon is currently very compact, without any connection complexity).

kgaillot commented on lib/common/options.c:
```diff
@@ -38,6 +38,445 @@ pcmk__cli_help(char cmd)
+         * be an ISO 8601 duration, which pcmk__valid_interval_spec() accepts.
+         */
+        PCMK_OPT_STONITH_WATCHDOG_TIMEOUT, NULL, "time", NULL,
+        "0", pcmk__valid_interval_spec,
```
Actually `pcmk__get_sbd_timeout()` parses it with `crm_get_msec()`. Interval specs are a superset, so I eventually want to get around to upgrading parsed timeouts to interval specs, but this one will be problematic because cluster and remote nodes pass it back and forth.

nrwahl2 replied:
That's my point in the TODO comment. `pcmk__valid_sbd_timeout()` parses it with `crm_get_msec()`, but the validation function (`pcmk__valid_interval_spec()`) accepts a superset of the allowed values.

Or are you just suggesting I update the comment to point to `pcmk__get_sbd_timeout()` instead of `pcmk__valid_sbd_timeout()` for clarity?

kgaillot replied:
Let's create a `pcmk__valid_timeout()` for `crm_get_msec()` values

nrwahl2 replied:
Good idea, but we'll need more than that. We'll need a new function that wraps `crm_get_msec()`, since `crm_get_msec()` uses a return value of -1 (`PCMK__PARSE_INT_DEFAULT`) to indicate an error. That's a valid value for `stonith-watchdog-timeout`.

Make the new function public for the same reason as `pcmk_parse_interval_spec()`?

kgaillot replied:
Sure

nrwahl2 replied:
It turns out literally everything is a valid stonith-watchdog-timeout in a sense... if it's negative or can't be parsed, it's treated the same as `-1` by `pcmk__valid_sbd_timeout()`. And even that might not be applied consistently in all contexts. I'll have to do some auditing of the various uses.

On the other hand, `crm_get_msec()` treats anything negative as invalid, and it's documented as requiring a nonnegative input. So the `-1` error return value doesn't create any ambiguity. It might be helpful to have a version that can handle negative numbers, I'm not sure just yet.

A negative timeout doesn't technically make sense, but it can have special meaning. So it might be useful to differentiate between passing a negative timeout value and passing a garbage value; right now, both produce an identical negative return code.

kgaillot replied:
Ouch, I forgot a negative value is valid for stonith-watchdog-timeout. Also, for compatibility reasons we don't want to reject anything that would have been accepted by the old validator, so that means we really can't do any validation until 3.0.0 (at which point a signed integer would be appropriate).

Our current definition of timeout (and duration) in Pacemaker Explained doesn't say anything about negative values. I believe stonith-watchdog-timeout is the only option that gives a special meaning to them.

nrwahl2 replied:
I just went through all the cases in both the validator function and the fencer. The good news is that the validator computes the same values as the fencer and passes when the value will work in the fencer (and vice-versa). Here's a list of cases.
```
Configured value:
* negative or invalid
  * If (SBD_WATCHDOG_TIMEOUT <= 0):
    * Value: 0
    * Validation: Pass
  * Else:
    * Value: 2 * SBD_WATCHDOG_TIMEOUT
    * Validation: Pass (unless multiplication overflows or sbd not running)
* 0 or unset
  * Value: 0
  * Validation: Pass
* positive
  * Value: <configured value>
  * Validation: Pass if (value >= SBD_WATCHDOG_TIMEOUT)
                (unless multiplication overflows or sbd not running)
                Otherwise fail
    * Note: This should be "pass if (value > SBD_WATCHDOG_TIMEOUT)" (strictly greater than)
```

The core problem is that we have two default values. In that sense, stonith-watchdog-timeout is poorly specified.
* If the configured value is absent, we default to "0".
* If the configured value is invalid (i.e., a garbage string), we default to "-1".
  * "-1" is equivalent to any negative number; we say "-1" for simplicity.

nrwahl2 replied:
> I believe stonith-watchdog-timeout is the only option that gives a special meaning to them.

That may be. It's a legitimate use case for stonith-watchdog-timeout (to indicate "auto"), since "0" is reserved for "disabled".

nrwahl2 replied:
It could also make sense to drop support for negative values altogether and have a choice between the string "auto" and a timeout. That doesn't fit as neatly into the "type" field of a table or metadata, but it's arguably clearer than "negative == auto".

kgaillot replied:
> The core problem is that we have two default values. In that sense, stonith-watchdog-timeout is poorly specified.
> 
>     * If the configured value is absent, we default to "0".
> 
>     * If the configured value is invalid (i.e., a garbage string), we default to "-1".
>       
>       * "-1" is equivalent to any negative number; we say "-1" for simplicity.
> 
> 
> I think it would make more sense to default to "0" (disabled) on invalid configured values. That raises compatibility problems.

Agreed, create a task for 3.0.0

> As it stands, the default in the doc table and the options array is wrong and can't be made correct.

We'll have to live with it until then

> It could also make sense to drop support for negative values altogether and have a choice between the string "auto" and a timeout. That doesn't fit as neatly into the "type" field of a table or metadata, but it's arguably clearer than "negative == auto".

That does make more sense, but we'd need an XSL transform for existing configs. Could be part of same task. I think it would be fine to document a type like `timeout or "auto"`. However given how much else needs to be done it's probably best just to leave it as negative.

nrwahl2 replied:
https://projects.clusterlabs.org/T749

kgaillot commented on lib/common/options.c:
```diff
@@ -947,13 +947,28 @@ add_desc(GString *s, const char *tag, const char *desc, const char *values,
+ * \param[in] option_list  Options whose metadata to format
+ * \param[in] len          Number of items in \p option_list
+ *
+ * \return A string containing OCF-compliant option metadata XML
```
Technically it's more OCF-like than OCF-compliant

nrwahl2 replied:
Ah yes... I don't see a lot of difference, but `<actions>` appears to be required by the OCF rng file, and we're using more `type` values (e.g., "time") than the OCF rng file allows.

kgaillot commented on include/crm/msg_xml_compat.h:
```diff
@@ -406,128 +406,8 @@ extern "C" {
-//! \deprecated Use \c PCMK_META_RECORD_PENDING instead
-#define XML_OP_ATTR_PENDING PCMK_META_RECORD_PENDING
+//! \deprecated Use \c PCMK_XE_PARAMETERS instead
+#define XML_TAG_PARAMS PCMK_XE_PARAMETERS
```
Looks like this accidentally deleted a whole bunch of stuff (probably in rebase)

nrwahl2 replied:
Oops

kgaillot reviewed (commented):
good start

kgaillot commented:
Once we have a new method, we can deprecate the "metadata" command-line option for daemons, and stop regression-testing them. However we do need a field for which daemon each option "belongs" to, so the deprecated options show only the relevant ones (so existing software that parses it remains reasonable). One change from the current output should be that each option should be associated with only one daemon (we can prefer the scheduler).

nrwahl2 commented:
> Once we have a new method, we can deprecate the "metadata" command-line option for daemons, and stop regression-testing them. However we do need a field for which daemon each option "belongs" to, so the deprecated options show only the relevant ones (so existing software that parses it remains reasonable).

That was my original plan but it shouldn't make any difference. There's no reason why any external tool should care which of Pacemaker's daemons process(es) a particular cluster option. All they should care about is that the option exists and how it affects **Pacemaker's** behavior. Anything else is a flawed design of the external tool IMO.

I don't want to downplay the current flawed design of our option output either, to be fair.

> One change from the current output should be that each option should be associated with only one daemon (we can prefer the scheduler).

Might as well, if pcs won't add add logic to filter out duplicates (RHEL-7702).

nrwahl2 commented:
Rebased to resolve conflicts. Also added three commits in the middle that just rename some watchdog timeout-related functions. Still have work to do on timeout validation.

kgaillot commented:
> > Once we have a new method, we can deprecate the "metadata" command-line option for daemons, and stop regression-testing them. However we do need a field for which daemon each option "belongs" to, so the deprecated options show only the relevant ones (so existing software that parses it remains reasonable).
> 
> That was my original plan but it shouldn't make any difference. There's no reason why any external tool should care which of Pacemaker's daemons process(es) a particular cluster option.

They don't, we forced them to. :)

> All they should care about is that the option exists and how it affects **Pacemaker's** behavior. Anything else is a flawed design of the external tool IMO.

All they really care about is a comprehensive list of options, which we didn't provide, so they had to ask all the daemons instead. But even that is not what they want since we duplicate some options.
 
> I don't want to downplay the current flawed design of our option output either, to be fair.
> 
> > One change from the current output should be that each option should be associated with only one daemon (we can prefer the scheduler).
> 
> Might as well, if pcs won't add add logic to filter out duplicates (RHEL-7702).

Yep, the goal is to continue working with existing software expecting the old style while giving them a real solution for the future.

nrwahl2 commented:
> They don't, we forced them to. :)

We forced them to query daemon metadata to get options. We didn't force anyone to care which daemon(s) include(s) which option, or whether multiple do.

Again, don't want to downplay the shortcomings of our current interface. It wasn't unreasonable to expect that there wouldn't be duplicates, anyway.

> Yep, the goal is to continue working with existing software expecting the old style while giving them a real solution for the future.

Makes sense.

nrwahl2 commented:
Rebased on current main. There are some new WIP commits that will be changed or removed, don't scrutinize :)

nrwahl2 commented on lib/common/options.c:
```diff
@@ -1041,6 +1059,11 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
         const char *opt_desc_long = option_list[lpc].description_long;
 
+        if ((filter != pcmk__opt_context_none)
+            && (filter != option_list[lpc].context)) {
```
We can switch this to `pcmk_is_set(option_list[lpc].context, filter)` in the future if any options have multiple context flags set. For now, an option holds a single enum value rather than `uint32_t` flag set., so we use this notation for clarity.

nrwahl2 commented:
Just start from the beginning. Enough has changed that that's easier than going through the comparison.

kgaillot reviewed (approved):
Beautiful

kgaillot commented:
Although it does need a rebase now

kgaillot commented:
Also, please add a commit for "make -C po update-po" so translators can work with the new material

nrwahl2 commented:
> Also, please add a commit for "make -C po update-po" so translators can work with the new material

I have no idea how gettext() and po files work, when this needs to be done, or why some strings are wrapped by `N_()` and others aren't -- but I'll gladly run the command you gave me.

nrwahl2 commented:
Rebased again to resolve new conflicts. Note that (as of the previous push) there's a PO commit.

kgaillot commented on daemons/based/pacemaker-based.c:
```diff
@@ -126,6 +126,21 @@ setup_stand_alone(GError **error)
+    gchar *s = pcmk__cluster_option_metadata(name, desc_short, desc_long,
+                                             pcmk__opt_context_based);
+
+    printf("%s", s);
```
Yep, we already have formatted output of XML (such as `crm_resource --show-metadata`) via `out->output_xml()`. Text output shows it as-is, XML output wraps it in a CDATA block.

Since we're defining the XML in this case (rather than passing on something from an agent), we don't need to use CDATA. In fact, in this case, text output shouldn't show XML at all. Text output can be a simple list, and XML output OCF-like.

If it winds up being trivial to use formatted output here after creating the functions, we might as well, but I wouldn't waste time specifically for it.

nrwahl2 replied:
> Yep, we already have formatted output of XML (such as crm_resource --show-metadata) via out->output_xml(). Text output shows it as-is, XML output wraps it in a CDATA block.

It's too late to change it, but I feel like it would've been better to transform it to a simple list for text output and show the XML as-is for XML output. In other words, my first inclination would've been to model this on `pcs resource describe`.

> Since we're defining the XML in this case (rather than passing on something from an agent), we don't need to use CDATA. In fact, in this case, text output shouldn't show XML at all. Text output can be a simple list, and XML output OCF-like.

Yep that was my plan. As mentioned above, it could be pcs-like. Although pcs may want to get XML output and parse it themselves, for reliability and ultimate control of formatting.

> If it winds up being trivial ...

I suspect it will (create a text object, use once and destroy). Could be wrong.

nrwahl2 replied:
> I feel like it would've been better to transform it to a simple list for text output

I suppose the counter-argument is that the default output should match that of `/path/to/agent meta-data`.

kgaillot replied:
> > Yep, we already have formatted output of XML (such as crm_resource --show-metadata) via out->output_xml(). Text output shows it as-is, XML output wraps it in a CDATA block.
> 
> It's too late to change it, but I feel like it would've been better to transform it to a simple list for text output and show the XML as-is for XML output. In other words, my first inclination would've been to model this on `pcs resource describe`.

It would be cool to have readable text output for agent metadata (like "Parameters: ...."). But the XML output will always need a CDATA block because we can't define an API schema for externally generated XML. Then again someone might want the raw XML metadata -- I suppose we could keep that as the default text output and finally make use of the `--text-fancy` option to show it in a readable way.
 
> > Since we're defining the XML in this case (rather than passing on something from an agent), we don't need to use CDATA. In fact, in this case, text output shouldn't show XML at all. Text output can be a simple list, and XML output OCF-like.
> 
> Yep that was my plan. As mentioned above, it could be pcs-like. Although pcs may want to get XML output and parse it themselves, for reliability and ultimate control of formatting.

Definitely, the whole point of XML output is that anyone parsing the output should use that.

nrwahl2 replied:
> But the XML output will always need a CDATA block because we can't define an API schema for externally generated XML.

OCF agents' metadata is *supposed* to be able to validate against the OCF RA schema (even though we don't provide it currently), right? I mean, tools like pcs implicitly rely on that already, and so would a conversion to text output.

It's a loose dependence, in that we can skip some things in the XML that are absent from the schema and vice-versa. And that might still be enough of a problem to preclude us from ever outputting the raw XML -- some XML in a third-party file that would fail to validate, would cause the Pacemaker output to fail to validate.

Though I'm not 100% sure that "Pacemaker output fails to validate when it's given input (the agent) that fails to validate" should be a deal-breaker.

As I finish typing this, I remember that fence agent metadata is another problem. No longdescs. Extra `<getopt>` elements. Etc. You know what? Nevermind.

> I suppose we could keep that as the default text output and finally make use of the --text-fancy option to show it in a readable way.

https://projects.clusterlabs.org/T750

kgaillot commented:
> > Also, please add a commit for "make -C po update-po" so translators can work with the new material
> 
> I have no idea how gettext() and po files work, when this needs to be done, or why some strings are wrapped by `N_()` and others aren't -- but I'll gladly run the command you gave me.

The make command regenerates translation files from the C source. The translation files just have each original string and its translation (if one exists).

Conceptually gettext uses the original string as a key and translations as values. Normally, `_()` (which is an alias for `gettext()`) is used to wrap a string to be translated at runtime, and gettext uses the original to get the translation. However that approach doesn't work for initializations of static objects. In those cases we have to mark the string with `N_()`, which actually does nothing, but gettext parses the markers to handle them specially. If an option description doesn't have `N_()`, that's just because it hasn't been translated yet -- it wouldn't hurt to use `N_()` for all of them.

nrwahl2 commented:
Deleted a curly brace during conflict resolution and forgot to build locally :( about to repush
